Throw an error if no where clause is given to Model.destroy()

ekmartin opened the pull request:
See #2847.

This throws an error similar to the one that's thrown if `Model.update()` is called with no where clause. I couldn't find a test that checks the same for `.update()` so I added one there too.

mickhansen commented on lib/model.js:
```diff
@@ -1412,6 +1412,13 @@ module.exports = (function() {
+    var self = this
+      , instances;
+
+    if (!options || !options.where) {
```
I belive we support a `truncate` parameter, if that is set it should't throw an error.

mickhansen replied:
Hmm or do we, can't find it right now, will have to check further.

mickhansen replied:
https://github.com/sequelize/sequelize/blob/master/lib/dialects/abstract/query-generator.js
https://github.com/sequelize/sequelize/blob/master/lib/dialects/postgres/query-generator.js#L405

We do support truncate down the line, so you might want to add a check and test to make sure that `truncate: true` will still work correctly.

ekmartin commented:
I updated it so `Model.destroy({ truncate: true })` doesn't throw now. I added a test for it, but it seems like there's a few problems with TRUNCATE on MSSQL (see Travis). I've changed the syntax from TRUNCATE to TRUNCATE TABLE, but there's still a problem with formatting the results in according to affected rows (which truncate doesn't return).

mickhansen commented:
@ekmartin perhaps change so that https://github.com/sequelize/sequelize/blob/master/lib/dialects/mssql/query.js#L152 atleast doesn't crash :) `data[0] && data[0].AFFECTEDROWS`. I'm not sure if there's a test for affected rows on truncate

mickhansen commented:
Thanks @ekmartin

bulkan commented on lib/model.js:
```diff
@@ -1412,6 +1412,13 @@ module.exports = (function() {
+    var self = this
+      , instances;
+
+    if (!options || !(options.where || options.truncate)) {
```
This condition fails if you just pass `Model.destroy({truncate: false})`

ekmartin replied:
`Model.destroy({ truncate: false })` shouldn't be allowed anyway? It's the same as `Model.destroy()`, which isn't allowed.

bulkan replied:
I was just surprised to have have to change all calls to `Model.destroy()` to `Model.destroy({where: {}})`

mickhansen replied:
Understandable @bulkan.
Actually one could argue that this if should only be in effect if `options` is defined.

If `destroy` is called with no options at all there's no chance that people are inadvertently using the old where syntax, it's pretty clear they want a full destroy.
